fe/fe-core/src/main/java/com/starrocks/consistency/PartitionVersionRecovery.java

```diff
@@ -0,0 +1,174 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.consistency;
+
+import com.starrocks.catalog.Database;
+import com.starrocks.catalog.LocalTablet;
+import com.starrocks.catalog.MaterializedIndex;
+import com.starrocks.catalog.OlapTable;
+import com.starrocks.catalog.Partition;
+import com.starrocks.catalog.PhysicalPartition;
+import com.starrocks.catalog.Table;
+import com.starrocks.catalog.Tablet;
+import com.starrocks.common.util.FrontendDaemon;
+import com.starrocks.common.util.concurrent.lock.LockType;
+import com.starrocks.common.util.concurrent.lock.Locker;
+import com.starrocks.persist.PartitionVersionRecoveryInfo;
+import com.starrocks.persist.PartitionVersionRecoveryInfo.PartitionVersion;
+import com.starrocks.server.GlobalStateMgr;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class PartitionVersionRecovery extends FrontendDaemon {
+    private static final Logger LOG = LogManager.getLogger(PartitionVersionRecovery.class);
+
+    public PartitionVersionRecovery() {
+        super("partition_version_recovery");
+    }
+
+    @Override
+    protected void runAfterCatalogReady() {
+        recover();
+    }
+
+    public void recover() {
+        GlobalStateMgr stateMgr = GlobalStateMgr.getCurrentState();
+        if (!stateMgr.getNodeMgr().getClusterInfo().checkTabletReportCacheUp(stateMgr.getDominationStartTimeMs())) {
+            LOG.warn("check tablet report time failed, wait for next round");
+            return;
+        }
+
+        List<PartitionVersion> partitionsToRecover = new ArrayList<>();
+        List<Long> dbIds = stateMgr.getLocalMetastore().getDbIds();
+        for (long dbId : dbIds) {
+            Database database = stateMgr.getDb(dbId);
+            if (database == null) {
+                continue;
+            }
+
+            if (database.isSystemDatabase()) {
+                continue;
+            }
+
+            Locker locker = new Locker();
+            locker.lockDatabase(database, LockType.READ);
+            try {
+                for (Table table : database.getTables()) {
+                    if (!table.isOlapTableOrMaterializedView()) {
+                        continue;
+                    }
+
+                    OlapTable olapTable = (OlapTable) table;
+
+                    for (Partition partition : olapTable.getAllPartitions()) {
+                        short replicaNum = olapTable.getPartitionInfo().getReplicationNum(partition.getId());
+
+                        PARTITION:
+                        for (PhysicalPartition physicalPartition : partition.getSubPartitions()) {
+                            long recoverVersion = -1L;
+                            for (MaterializedIndex idx : physicalPartition.getMaterializedIndices(
+                                    MaterializedIndex.IndexExtState.VISIBLE)) {
```

Comment:
why only VISIBLE?

Reply:
Shadow index is an intermediate state for schema change. The tablets of shadow index maybe unhealthy at certain stages of schema change.